### PR TITLE
Completed migration of Environment

### PR DESCRIPTION
**Pull request by dylannakahodo**
I finished migrating the Environment code from the old-version-Spring-2017 branch.
- Function definitions are contained in the source file
- Only function declarations are in the header file

I discussed with Austin regarding `contains()` in Environment, and we decided that it will only check for the key in the current Environment, while `is_defined()` will recursively check the current Environment and all parents.

I also added the unit-tests for Environment and updated documentation.

**I think everything should be correct. Please let me know if you notice anything wrong with my implementation.**

### CODE REVIEW

File: tst/shaka_scheme/system/base/unit-Environment.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  Symbol key("x");
+
+  // When: you try to get a key with no binding in the environment
+  // Then: an InvalidInputException is thrown
+  ASSERT_THROW(e.get_value(key), InvalidInputException);
+
+}
+
+TEST(EnvironmentUnitTest, contains) {
+
```
Comment (btwooton) [inline tst/shaka_scheme/system/base/unit-Environment.cpp]: No test case for contains? Looking at the code, I'm sure it works, but may be good to include this test as well just for sake of coverage.

**Review by btwooton — APPROVED** [review]
Everything looks solid to me. Nice pull request Dylan. Comments in the Environment.hpp file are well formed and descriptive. All method implementations look sound. Tests look good, albeit you may have forgot the unit test for `contains`. I approve merging of these changes. Well done.

File: tst/shaka_scheme/system/base/unit-Environment.cpp (hunk 1 above)
Comment (dylannakahodo) [reply tst/shaka_scheme/system/base/unit-Environment.cpp]: Whoops, yeah that should have a test case. I'll add that and commit.

File: tst/shaka_scheme/system/base/unit-Environment.cpp [hunk 2]
```diff
@@ -109,7 +109,22 @@ TEST(EnvironmentUnitTest, get_value_invalid_key_failure) {
 }
 
 TEST(EnvironmentUnitTest, contains) {
+  // Given: an environment and different symbols and values
+  Environment e(nullptr);
+  Symbol key1("x");
+  Symbol key2("y");
+  auto x_value = create_node(String("x test"));
```
Comment (dylannakahodo) [inline tst/shaka_scheme/system/base/unit-Environment.cpp]: I added the unit test for contains. Thanks for catching that

Comment (btwooton) [reply tst/shaka_scheme/system/base/unit-Environment.cpp]: No problem. Thanks for the commit. I'll let Austin do the rebase and merge.

File: CMakeLists.txt [hunk 3]
```diff
@@ -49,7 +49,7 @@ set(SOURCE_FILES
         src/shaka_scheme/system/base/String.cpp
         src/shaka_scheme/system/base/Symbol.cpp
         src/shaka_scheme/system/base/DataPair.cpp
-        src/shaka_scheme/system/exceptions/MissingImplementationException.hpp)
+        src/shaka_scheme/system/exceptions/MissingImplementationException.hpp src/shaka_scheme/system/base/Environment.cpp src/shaka_scheme/system/base/Environment.hpp)
```
Comment (CinchBlue) [inline CMakeLists.txt]: Remove `src/shaka_scheme/system/base/Environment.hpp`. Headers should not be in the source files.

File: src/shaka_scheme/system/base/Environment.hpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+   * @param key They key to be used for lookup in the environment
+   * @param data The data associated with the key
+   */
+  void set_value(const Key& key, Value data) override;
+
+  /**
+   * @brief Returns a reference to the value referred to by the key. Key
+   *        must exist, otherwise, it throws a runtime error.
+   * @param key The key to lookup and return the value for.
+   * @return The value associated with the given key
```
Comment (CinchBlue) [inline src/shaka_scheme/system/base/Environment.hpp:59]: There should be another `@throws InvalidInputException when the key does not exist in the environment bindings`. The function should throw an `InvalidInputException`.

File: tst/shaka_scheme/system/base/unit-Environment.cpp [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+  Symbol key2("y");
+  auto value2 = create_node(String("parent test"));
+
+  // When: you set the key to the specified value in only the parent
+  parent->set_value(key2, value2);
+
+  // Then: the correct value will be returned
+  ASSERT_EQ(e.get_value(key2), value2);
+}
+TEST(EnvironmentUnitTest, get_value_invalid_key_failure) {
```
Comment (CinchBlue) [inline tst/shaka_scheme/system/base/unit-Environment.cpp]: I see that you already changed the implementation correctly if the unit test case passes. Therefore, just change the documentation on the function.

File: tst/shaka_scheme/system/base/unit-Environment.cpp [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+// Created by dylan on 9/20/17.
+//
+
+#include <gmock/gmock.h>
+#include "shaka_scheme/system/base/Environment.hpp"
+#include "shaka_scheme/system/base/Data.hpp"
+
+using namespace shaka;
+
+TEST(EnvironmentUnitTest, parent_constructor) {
```
Comment (CinchBlue) [inline tst/shaka_scheme/system/base/unit-Environment.cpp]: Please put Doxygen comments with `@brief` one-line descriptions of the form `@brief Test: blah blah blah` etc. Look at `unit-DataNode.cpp` for an example.

**Review by CinchBlue — CHANGES_REQUESTED** [review]
Hi -- very good first pull request. It's very clean. There's just a few holes in the documentation/commenting that need to be fixed before this goes out.

Just make an additional commit to this pull-request branch -- it should automatically update here once you push it to your own branch.

**Comment by dylannakahodo** [comment]
Finished documentation and comments that Austin had.

**Review by CinchBlue — APPROVED** [review]

**Comment by CinchBlue** [comment]
Looks like the changes have all been addressed. I approve this pull request.

**Pull request merged**